class.c

```diff
@@ -805,14 +805,80 @@ Perl_class_seal_stash(pTHX_ HV *stash)
         aux->xhv_class_initfields_cv = initfields;
     }
     else {
-        /* we had errors, clean up and don't populate initfields */
+        /* we had errors, clean up */
+
+        SvREFCNT_dec(aux->xhv_class_superclass);
+        aux->xhv_class_superclass = NULL;
+
+        /* clean up adjust blocks */
+        SvREFCNT_dec(aux->xhv_class_adjust_blocks);
+        aux->xhv_class_adjust_blocks = NULL;
+
+        /* name to slot index */
+        SvREFCNT_dec(aux->xhv_class_param_map);
+        aux->xhv_class_param_map = NULL;
+
+        /* clean up the ops for defaults for fields, if any, since
+           padname_free() doesn't.
+         */
         PADNAMELIST *fieldnames = aux->xhv_class_fields;
         if (fieldnames) {
             for(SSize_t i = PadnamelistMAX(fieldnames); i >= 0 ; i--) {
                 PADNAME *pn = PadnamelistARRAY(fieldnames)[i];
                 op_free(PadnameFIELDINFO(pn)->defop);
+                PadnameFIELDINFO(pn)->defop = NULL;
             }
+            PadnamelistREFCNT_dec(fieldnames);
+            aux->xhv_class_fields = NULL;
+        }
+
+        /* TODO: cleanup DOES if added above */
```

Comment:
At first, I was confused by this comment. (What's so special about a `DOES` method that is not handled by the generic cleanup below? Why not add the code right now instead of relying on some unknown future developer? What are the consequences of not cleaning up `DOES`?)

Turns out this comment is entirely appropriate because the code that would create a `DOES` method only exists in the form of a "TODO" comment, too. There really is nothing to do here until `DOES` is actually implemented.

Reply:
I thought about it and removed the comment.

If DOES() is synthesized before class_seal_stash() then we'll get extra warnings in the tests this change adds.

If DOES() is synthesized in class_seal_stash() then it won't exist at this point.

I'm going to make some other changes too, I think it's reasonable to put the clean up in a new function to keep class_seal_stash() focused.